apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/partition/SimpleRollingPartitioner.java

```diff
@@ -18,16 +18,18 @@
 
 package org.apache.skywalking.apm.commons.datacarrier.partition;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 /**
  * use normal int to rolling.
  */
 public class SimpleRollingPartitioner<T> implements IDataPartitioner<T> {
-    @SuppressWarnings("NonAtomicVolatileUpdate")
-    private volatile int i = 0;
+
+    private final AtomicInteger i = new AtomicInteger();
```

Comment:
Let's talk about this. I know your point of bug, it would not increase every time. But it has higher performance with no lock, in the partition, there is a lock to avoid the real issue. So this is actually not a bug in the real world.
This is another typical thing, codes are not by the book.

Reply:
> in the partition, there is a lock to avoid the real issue.

Could you tell me where the lock is? I want to take a look.

Reply:
Channel index. There is a atomic integer as index pointer.